src/reader/async_tokio.rs

```diff
@@ -0,0 +1,462 @@
+//! This is an implementation of [`Reader`] for reading from a [`AsyncBufRead`]
+//! as underlying byte stream. This reader fully implements async/await so reading
+//! can use non-blocking I/O.
+
+use std::future::Future;
+use std::ops::{Deref, DerefMut};
+use std::pin::Pin;
+use tokio::io::{self, AsyncBufRead, AsyncBufReadExt};
+
+use crate::events::Event;
+use crate::name::{QName, ResolveResult};
+use crate::reader::buffered_reader::impl_buffered_source;
+use crate::reader::{is_whitespace, BangType, NsReader, ParseState, ReadElementState, Reader};
+use crate::{Error, Result};
+
+/// A struct that adapts any [`AsyncBufRead`] reader to read an XML.
+/// See [`Reader::read_event_into_async`] for details.
+#[derive(Debug, Clone)]
+pub struct TokioAdapter<R>(pub R);
```

Comment:
This `Reader` can't really override the `into_inner` implementation of `impl<R> Reader<R>` in `reader/mod.rs`, so a `Reader<TokioAdapter<R>>` only returns the `TokioAdapter<R>` when calling `into_inner`. To get the `AsyncBufRead` back out, `TokioAdapter` should also implement an `into_inner`.

Maybe there's a better way to do this like I did in #425 with the `InnerReader` trait?

Reply:
`TokioAdapter` has a public field. Anyway, I think I found a better way and be able to hide `TokioAdapter` as an implementation detail. I'll update PR tomorrow